msal/application.py

```diff
@@ -69,6 +70,36 @@ def _clean_up(result):
     return result
 
 
+def _preferred_browser():
+    """Register Edge and return a name suitable for webbrowser.get(...),
+    if running on Linux and there is no BROWSER env var, otherwise return None.
+    """
+    browser_path = "/usr/bin/microsoft-edge"  # Use a full path owned by sys admin
+    user_has_no_preference = "BROWSER" not in os.environ
+    user_opted_in_for_edge = "microsoft-edge" in os.environ.get("BROWSER", "")
+    if ((user_opted_in_for_edge or user_has_no_preference)
+            and sys.platform == "linux"  # On Linux, only Edge will have CA support
+            and os.path.exists(browser_path)):  # Edge is usually installed here
+        try:
+            import webbrowser  # Lazy import. Some distro may not have this.
+            browser_name = "msal-edge"  # A unique name rather than "microsoft-edge"
+                # otherwise `BROWSER="microsoft-edge"; webbrowser.get("microsoft-edge")`
+                # would return a GenericBrowser instance which won't work.
+            try:
+                registration_available = isinstance(
+                    webbrowser.get(browser_name), webbrowser.BackgroundBrowser)
+            except webbrowser.Error:
+                registration_available = False
+            if not registration_available:
+                logger.debug("Register %s with %s", browser_name, browser_path)
+                webbrowser.register(  # Even double-register happens to work fine
+                    browser_name, None, webbrowser.BackgroundBrowser(browser_path))
```

Comment:
Off topic: I actually find the implementation of `webbrowser.BackgroundBrowser` problematic.

https://github.com/python/cpython/blob/2b496e79293a8b80e8ba0e514e186b3b1467b64b/Lib/webbrowser.py#L190-L202

```py
class BackgroundBrowser(GenericBrowser):
    def open(self, url, new=0, autoraise=True):
        ...
            if sys.platform[:3] == 'win':
                p = subprocess.Popen(cmdline)
            ...
            return (p.poll() is None)
```

Per https://docs.python.org/3/library/subprocess.html#subprocess.Popen.poll:

> `Popen.poll()`
> Check if child process has terminated. Set and return returncode attribute. Otherwise, returns None.

Say my computer is super fast and `p.poll()` immediately returns `0`, this `open` method will return `False`. Then MSAL will think the browser is not opened, showing the warning 

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/c687d5b0a198c11c2ff99fd0ca9feaf73a7376f0/msal/oauth2cli/authcode.py#L210-L221

We can mimic this behavior by putting a `time.sleep(2)` before `return (p.poll() is None)` to slow down `webbrowser.BackgroundBrowser.open`.

Therefore, it should actually be 

```py
poll_result = p.poll()
return poll_result is None or poll_result == 0
```

Perhaps we can submit a PR to https://github.com/python/cpython

Reply:
I tested on Windows. We need to explicitly request a web browser, like `firefox` (it should be on `PATH`).

Reply:
> I tested on Windows. We need to explicitly request a web browser, like `firefox` (it should be on `PATH`).

@jiasli Did you actually see any abnormal behavior when you were testing it on Windows? [Line 81 in the current PR](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/388/files#diff-4d32d284fa6acf43894719eff11cb366e5969e0b61593bfd00c68a4cf9c441caR81) would at least make this PR's behavior would only occur on Linux. I tested it on Windows just now, and its behavior is same as before, just opening my default browser.

> Off topic: I actually find the implementation of `webbrowser.BackgroundBrowser` problematic.
> 
> https://github.com/python/cpython/blob/2b496e79293a8b80e8ba0e514e186b3b1467b64b/Lib/webbrowser.py#L190-L202
> 
> ```python
> class BackgroundBrowser(GenericBrowser):
>     def open(self, url, new=0, autoraise=True):
>         ...
>             if sys.platform[:3] == 'win':
>                 p = subprocess.Popen(cmdline)
>             ...
>             return (p.poll() is None)
> ```
> 
> Per https://docs.python.org/3/library/subprocess.html#subprocess.Popen.poll:
> 
> > `Popen.poll()`
> > Check if child process has terminated. Set and return returncode attribute. Otherwise, returns None.
> 
> Say my computer is super fast and `p.poll()` immediately returns `0`, this `open` method will return `False`. Then MSAL will think the browser is not opened, showing the warning

In theory, yes, `subprocess.Popen(cmdline).poll()` could have finished very fast. (Actually I was not aware of this detail. Thanks for teaching me this new knowledge.)

But in our case, the browser is meant to open for user interaction. Even if the user would have a super fast computer, the opened browser window would not be immediately terminated unless the user closes it immediately. Even if the user does so, that subsequent MSAL warning message would seem quite reasonably in such a corner case.

Unless I haven't understood the problem thoroughly. Please let me know if I miss something.

Reply:
> Did you actually see any abnormal behavior when you were testing it on Windows?

In reality, no. This issue should only be a corner case.

> the opened browser window would not be immediately terminated unless the user closes it immediately.

This is not true per my observation. On Windows, even if the browser is not closed, `p.poll()` will return `0` very quickly. Putting a `time.sleep(2)` before `p.poll()` can easily repro this. This delay can also happen when the execution of `webbrowser` is suspended for some reason, for example, when garbage collection (GC) kicks in, CPU goes to 100%, etc.

Reply:
> > > I tested on Windows. We need to explicitly request a web browser, like `firefox` (it should be on `PATH`).
> >
> > @jiasli Did you actually see any abnormal behavior when you were testing it on Windows?
>
> In reality, no. This issue should only be a corner case.

I thought you were seeing some symptom that could be fix by your very specific suggestion ("need to explicitly request a web browser, like `firefox`"). If that is not the case, then it is not actionable here.

You did raise some good points on Windows. I think you can create yet another issue for the "acquire_token_interactive() on Windows" topic, and then we can keep track there. The current PR here contains new behaviors which are [only invoked when running on Linux](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/388/files#diff-4d32d284fa6acf43894719eff11cb366e5969e0b61593bfd00c68a4cf9c441caR81).

Reply:
I know, that's why I marked it as "off topic" at the very beginning. 😆